File: Algorithm/leetcode_1248.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_1248.py
# Author            : Andy
# Date              : 30.07.2022
# Last Modified Date: 30.07.2022
# Last Modified By  : Andy


class Solution:
    def numberOfSubarrays(self, nums, k):
        dic = {0:1}
        count, sums = 0, 0
        for i in range(len(nums)):
            nums[i] = nums[i] % 2
            sums += nums[i]
            if sums -k in dic:
                count += dic[sums-k]
            dic[sums] = dic.get(sums, 0) + 1
        return count

File: Algorithm/leetcode_1248.go
/**
 * File              : leetcode_1248.go
 * Author            : Andy
 * Date              : 30.07.2022
 * Last Modified Date: 30.07.2022
 * Last Modified By  : Andy
 */

// 这题与前面560，930不同之处在于需要对数据进行预先处理，将奇数转成1，偶数转成0
// 当然这个题也可以出成偶数的，道理是一样的

func numberOfSubarrays(nums []int, k int) int{
    count, sum := 0, 0
    hash := map[int]int{0:1}
    for i:=0;i<len(nums); i++{
        nums[i] = nums[i] % 2
        sum += nums[i]
        if hash[sum-k] > 0 {
            count += hash[sum-k]
        }
        hash[sum] ++
    }
    return count
}
